gpu_tests/test_hub_utils.py

```diff
@@ -0,0 +1,233 @@
+        "best_of": 1,
+        "echo": True,
+        "logprobs": 0,
+        "seed": 1,
+    }
+
+    generated_text = generator.generate(**request_object)
+
+    # Destroy torch distributed process groups. This needs to be executed in each spawned process
+    # https://github.com/pytorch/pytorch/issues/48203
+    dist.destroy_process_group()
+    return generated_text
+
+
+def generate_using_generator_hub_interface(cfg: MetaseqConfig, **kwargs):
+    model_path = os.path.join(os.path.dirname(__file__), "125m")
+    kwargs = {
+        "suffix": f"-model_part-{torch.distributed.get_rank()}",
+        "criterion": "cross_entropy",
+        "bpe": "hf_byte_bpe",
+        "post_build_model_hook": default_post_build_model_hook,
+        "bpe_merges": os.path.join(kwargs["model_path"], "gpt2-merges.txt"),
+        "merges_filename": os.path.join(kwargs["model_path"], "gpt2-merges.txt"),
+        "bpe_vocab": os.path.join(kwargs["model_path"], "gpt2-vocab.json"),
+        "vocab_filename": os.path.join(kwargs["model_path"], "gpt2-vocab.json"),
+        "bpe_add_prefix_space": False,
+        "specify_arch": True,
+        "tensor_parallel_init_model_on_gpu": True,
+        # 'batch_size': None,
+        # 'batch_size_valid': None
+    }
+
+    # Need to enable this context otherwise fsdp_wrap is a no op. See fairscale.nn.wrap
+    with fsdp_enable_wrap(cfg.distributed_training, use_sharded_state=True):
+        x = hub_utils.from_pretrained(
+            model_path,
+            "reshard.pt",
+            **kwargs,
+        )
+
+    generator = GeneratorHubInterface(
+        x["args"],
+        x["task"],
+        x["models"],
+        moe_disable_padding=False,
+        skip_prepare_for_inference=True,
+    )
+
+    override_task(generator)
+    generator = generator.cuda()
+
+    tokenized_sentences = [torch.tensor(PROMPT)]
+    hypothesis = generator.generate(
+        tokenized_sentences=tokenized_sentences,
+        score_reference=True,
+        batch_size=1,
+        compute_vocab_dist=False,
+    )
+
+    # Destroy torch distributed process groups. This needs to be executed in each spawned process
+    # https://github.com/pytorch/pytorch/issues/48203
+    dist.destroy_process_group()
+    return hypothesis
+
+
+class TestHubUtils(unittest.TestCase):
+    def test_generator_interface(self):
+        model_path = os.path.join(os.path.dirname(__file__), "125m")
+        cfg = create_generation_config_with_defaults(
+            model_path, ddp_backend="fully_sharded"
+        )
+
+        overall_generation = distributed_utils.call_main(
+            cfg, generate_using_generator_interface
+        )
+        # We can potentially pass a list of inputs to the generate function.
+        # Here, we look at the first (and only) generation
+        [generation_for_prompt] = overall_generation
+        # We use best_of = 1, so we get only one beam search result
+        [generated_beam] = generation_for_prompt
+
+        EXPECTED_OUTPUT = {
```

Comment:
Should we consider using a regression testing framework to facilitate updating the expected output in the future?

https://pypi.org/project/pytest-regressions/

Reply:
Good point! Looking into this

Reply:
@stephenroller  Made this change